src/merkle/tiered_smt/storage.rs

```diff
@@ -0,0 +1,94 @@
+use super::{BTreeMap, ContentType, Leaf, NodeIndex, Vec, Word};
+use core::{cell::RefCell, fmt};
+
+#[derive(Debug, Default)]
+pub struct Storage {
+    leaves: RefCell<BTreeMap<NodeIndex, Vec<Leaf>>>,
+    nodes: RefCell<BTreeMap<NodeIndex, Word>>,
+    types: RefCell<BTreeMap<NodeIndex, ContentType>>,
+}
+
+impl Storage {
+    pub fn get_leaves(&self, index: &NodeIndex) -> Result<Option<Vec<Leaf>>, StorageError> {
+        self.leaves
+            .try_borrow()
+            .map_err(|_| StorageError::PoisonedInternalWrapper)
+            .map(|leaves| leaves.get(index).cloned())
+    }
+
+    pub fn get_node(&self, index: &NodeIndex) -> Result<Option<Word>, StorageError> {
+        self.nodes
+            .try_borrow()
+            .map_err(|_| StorageError::PoisonedInternalWrapper)
+            .map(|nodes| nodes.get(index).copied())
+    }
+
+    pub fn get_type(&self, index: &NodeIndex) -> Result<Option<ContentType>, StorageError> {
+        self.types
+            .try_borrow()
+            .map_err(|_| StorageError::PoisonedInternalWrapper)
+            .map(|types| types.get(index).copied())
+    }
```

Comment:
The idea is to abstract any logic from the storage. It will be basically a data repository that will `set/fetch`, and all logic is delegated to the concrete implementation of the tree. There is a function there that will unwrap a `None` to empty for the type of the storage.

This is noop during compilation, and makes maintenance easier as any algorithm update will be done only on the tiered merkle tree code.